specifications/android.md

```diff
@@ -82,32 +91,43 @@ There is a third action the connector SHOULD handle:
 ### org.unifiedpush.android.connector.NEW_ENDPOINT
 
 The distributor MUST send this action to the registered application to confirm the registration of an end user application, when a registered application send again an action with the intent org.unifiedpush.android.distributor.REGISTER and a valid token, or when the endpoint change with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* endpoint: the endpoint
+* token (String): the token supplied by the end user application during registration
+* endpoint (String): the endpoint
 
 ### org.unifiedpush.android.connector.REGISTRATION_FAILED
 
-The distributor MUST send this action to the registered application if the token is already registered for another application or if the registration can not be processed (for instance when the distributor is not connected to its provider server) with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* message: this extra MAY be sent to gives an error message
+The distributor MUST send this action to the registered application if:
+* the token is already registered for another application
+* the registration can not be processed (for instance when the distributor is not connected to its provider server)
+* a requested feature is not supported by the distributor
+
+The action contains the 2 following extras:
+* token (String): the token supplied by the end user application during registration
+* message (String): this extra MAY be sent to gives an error message
 
 The connector MUST change the connection token received with this action for the next registration.
 
 ### org.unifiedpush.android.connector.MESSAGE
 
 The distributor MUST send this action to the registered application to forward a push message received from the provider to the end user application.
 
-It MUST be send with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* message: the push message sent by the application server. It MUST be the raw POST data received by the rewrite proxy.
+It MUST be send with the following extra:
+* token (String): the token supplied by the end user application during registration
+
+It MUST be send with at least one of the 2 following extras:
+* message (String): the push message sent by the application server, as a string. It MUST be set if the BYTES_MESSAGE feature was not requested during registration. If set, it MUST be the raw POST data received by the rewrite proxy.
+* bytesMessage (ByteArray): the push message sent by the application server, as an array of bytes. It MAY be ommited. If set, it MUST be the raw POST data received by the rewrite proxy.
+
+The distributor MAY set message on top of bytesMessage if BYTES_MESSAGE feature was requested.
 
```

Comment:
While correct, the paragraph is hard to parse. May I suggest something like this:

```
If the BYTES_MESSAGE feature was requested, it MUST send the following extras:
* token (String): the token supplied by the end user application during registration
* bytesMessage (ByteArray): the push message sent by the application server, as an array of bytes.

If the BYTES_MESSAGE feature was requested, it MAY additionally send the message as a string:
* message (String): the push message sent by the application server, as a string.

If the BYTES_MESSAGE feature was not requested, it MUST send the following extras:
* token (String): the token supplied by the end user application during registration
* message (String): the push message sent by the application server, as a string.
```

Reply:
We should add the message (String) is not the full raw POST data. It works only if the message contains characters with char code between 0 and 127

Reply:
I've kept the req about RAW post data for the bytesMessage use case, and put nothing for the string version.